src/audio/multiband_drc/multiband_drc_generic.c

```diff
@@ -0,0 +1,411 @@
+			mix_out = sat_int32((int64_t)mix_out + *buf_src_band);
+			buf_src_band += PLATFORM_MAX_CHANNELS;
+		}
+
+		if (enable_deemp)
+			*buf_sink = iir_df2t_2stage_biquads(deemp_s, mix_out);
+		else
+			*buf_sink = mix_out;
+
+		buf_src++;
+		buf_sink++;
+	}
+}
+
+ /* This graph illustrates the buffers declared in the following default functions, as the example
+  * of a 3-band Multiband DRC:
+  *
+  *            :buf_src[nch]                            :buf_drc_sink[nch*nband]
+  *            :                                        :
+  *            :                           o-[]-> DRC0 -[]--o
+  *            :                           | :          :   |
+  *            :                 3-WAY     | :          :   |
+  *    source -[]-> EQ EMP --> CROSSOVER --o-[]-> DRC1 -[]-(+)--> EQ DEEMP -[]-> sink
+  *                                        | :          :   |               :
+  *                                        | :          :   |               :
+  *                                        o-[]-> DRC2 -[]--o               :
+  *                                          :                              :
+  *                                          :buf_drc_src[nch*nband]        :buf_sink[nch]
+  */
+#if CONFIG_FORMAT_S16LE
+static void multiband_drc_s16_default(const struct comp_dev *dev,
+				      const struct audio_stream *source,
+				      struct audio_stream *sink,
+				      uint32_t frames)
+{
+	struct multiband_drc_comp_data *cd = comp_get_drvdata(dev);
+	struct multiband_drc_state *state = &cd->state;
+	int32_t buf_src[PLATFORM_MAX_CHANNELS];
+	int32_t buf_sink[PLATFORM_MAX_CHANNELS];
+	int32_t buf_drc_src[PLATFORM_MAX_CHANNELS * SOF_MULTIBAND_DRC_MAX_BANDS];
+	int32_t buf_drc_sink[PLATFORM_MAX_CHANNELS * SOF_MULTIBAND_DRC_MAX_BANDS];
+	int32_t *band_buf_drc_src;
+	int32_t *band_buf_drc_sink;
+	int16_t *x;
+	int16_t *y;
+	int idx_src = 0;
+	int idx_sink = 0;
+	int ch;
+	int band;
+	int i;
+	int nch = source->channels;
+	int nband = cd->config->num_bands;
+	int enable_emp_deemp = cd->config->enable_emp_deemp;
+
+	for (i = 0; i < frames; ++i) {
+		for (ch = 0; ch < nch; ch++) {
+			x = audio_stream_read_frag_s16(source, idx_src);
+			buf_src[ch] = *x << 16;
+			idx_src++;
+		}
+
+		band_buf_drc_src = buf_drc_src;
+		band_buf_drc_sink = buf_drc_sink;
+		for (band = 0; band < nband; ++band) {
+			if (band == 0) {
+				multiband_drc_process_emp_crossover(state, cd->crossover_split,
+								    buf_src, buf_drc_src,
+								    enable_emp_deemp, nch, nband);
+			}
+
+			multiband_drc_process_drc(&state->drc[band], &cd->config->drc_coef[band],
+						  band_buf_drc_src, band_buf_drc_sink, 2, nch);
+			band_buf_drc_src += PLATFORM_MAX_CHANNELS;
+			band_buf_drc_sink += PLATFORM_MAX_CHANNELS;
+
+			if (band == nband - 1) {
+				multiband_drc_process_deemp(state, buf_drc_sink, buf_sink,
+							    enable_emp_deemp, nch, nband);
+			}
+		}
+
+		for (ch = 0; ch < nch; ch++) {
+			y = audio_stream_read_frag_s16(sink, idx_sink);
```

Comment:
Use audio_stream_write_frag_s16(). The macros for read and write are not the same.